opm/simulators/utils/MPIPacker.cpp

```diff
@@ -19,343 +19,118 @@
-typename std::enable_if<I != std::tuple_size<Tuple>::value, void>::type
-pack_tuple_entry(const Tuple& tuple, std::vector<char>& buffer,
-                 int& position, Opm::Parallel::MPIComm comm)
-{
-    pack(std::get<I>(tuple), buffer, position, comm);
-    pack_tuple_entry<I+1>(tuple, buffer, position, comm);
-}
-
-template<class... Ts>
-void pack(const std::tuple<Ts...>& data, std::vector<char>& buffer,
-          int& position, Opm::Parallel::MPIComm comm)
-{
-    pack_tuple_entry(data, buffer, position, comm);
-}
-
-void pack(const char* str, std::vector<char>& buffer, int& position,
-          Opm::Parallel::MPIComm comm)
-{
-#if HAVE_MPI
-    std::size_t length = strlen(str)+1;
+    std::size_t length = data.size();
     MPI_Pack(&length, 1, Dune::MPITraits<std::size_t>::getType(), buffer.data(),
-        buffer.size(), &position, comm);
-    MPI_Pack(str, strlen(str)+1, MPI_CHAR, buffer.data(), buffer.size(),
-         &position, comm);
-#else
-    (void) str;
-    (void) comm;
-    (void) buffer;
-    (void) position;
-#endif
-}
-
-void pack(const std::string& str, std::vector<char>& buffer, int& position,
-          Opm::Parallel::MPIComm comm)
-{
-    pack(str.c_str(), buffer, position, comm);
-}
-
-template<std::size_t Size>
-void pack(const std::bitset<Size>& data, std::vector<char>& buffer,
-          int& position, Opm::Parallel::MPIComm comm)
-{
-    Packing<std::bitset<Size>>::pack(data, buffer, position, comm);
-}
-
-void pack(const Opm::time_point& data, std::vector<char>& buffer, int& position,
-          Opm::Parallel::MPIComm comm)
-{
-    pack(Opm::TimeService::to_time_t(data), buffer, position, comm);
-}
-
-
-/// Mpi::unpack routines
-
-template<class T>
-void unpack(T*, const std::size_t&, std::vector<char>&, int&,
-            Opm::Parallel::MPIComm, std::integral_constant<bool, false>)
-{
-    OPM_THROW(std::logic_error, "Packing not (yet) supported for this non-pod type.");
-}
-
-template<class T>
-void unpack(T* data, const std::size_t& l, std::vector<char>& buffer, int& position,
-            Opm::Parallel::MPIComm comm,
-            std::integral_constant<bool, true>)
-{
-#if HAVE_MPI
-    MPI_Unpack(buffer.data(), buffer.size(), &position, data, l,
-               Dune::MPITraits<T>::getType(), comm);
-#else
-    (void) data;
-    (void) comm;
-    (void) l;
-    (void) buffer;
-    (void) position;
-#endif
-}
-
-template<class T>
-void unpack(T* data, const std::size_t& l, std::vector<char>& buffer, int& position,
-            Opm::Parallel::MPIComm comm)
-{
-    unpack(data, l, buffer, position, comm, typename std::is_pod<T>::type());
-}
-
-template<std::size_t I = 0, typename Tuple>
-typename std::enable_if<I == std::tuple_size<Tuple>::value, void>::type
-unpack_tuple_entry(Tuple&, std::vector<char>&, int&,
-                   Opm::Parallel::MPIComm)
-{
+             buffer.size(), &position, comm);
+    MPI_Pack(data.c_str(), length, MPI_CHAR, buffer.data(), buffer.size(),
+             &position, comm);
```

Comment:
Aren't you make an assumption here that might be violated for some strings?

Previously we extracted the C string representation that is equivalent to the std::string (even for multibyte encodings). Now we are assuming that this is not a multibyte string (i.e. strleng(data.c_str()==data.size() holds).